File: src/java/magma/compile/rule/InfixRule.java
package magma.compile.rule;

import magma.compile.CompileError;
import magma.compile.Node;
import magma.compile.context.StringContext;
import magma.result.Err;
import magma.result.Result;

public record InfixRule(Rule left, String infix, Rule right) implements Rule {
    @Override
    public Result<Node, CompileError> parse(String input) {
        int index = input.indexOf(infix);
        if (index >= 0) {
            String left = input.substring(0, index);
            String right = input.substring(index + infix.length());
            return this.left.parse(left).and(() -> this.right.parse(right)).mapValue(tuple -> tuple.left().merge(tuple.right()));
        } else {
            return new Err<>(new CompileError("Infix '" + infix + "' not present", new StringContext(input)));
        }
    }

    @Override
    public Result<String, CompileError> generate(Node node) {
        return left.generate(node).and(() -> right.generate(node)).mapValue(tuple -> tuple.left() + infix + tuple.right());
    }
}

File: src/clang/magma/compile/rule/text/InfixRule.c
#include "InfixRule.h"
int __lambda0__(){return right;
}
int __lambda1__(){return tuple;
}
magma.compile.rule.text.public InfixRule(magma.compile.rule.Rule left, String infix, magma.compile.rule.Rule right, magma.compile.rule.locate.Locator locator){this.left = left;
        this.infix = infix;
        this.right = right;
        this.locator = locator;
}
magma.result.Result<magma.compile.Node, magma.compile.CompileError> parse(String input){return locator.locate(input, infix).map(index -> {
            String left = input.substring(0, index);
            String right = input.substring(index + infix.length());
            return this.left.parse(left).and(() -> this.right.parse(right)).mapValue(tuple -> tuple.left().merge(tuple.right()));
        }).orElseGet(() -> {
            return new Err<>(new CompileError("Infix '" + infix + "' not present", new StringContext(input)));
        });
}
magma.result.Result<String, magma.compile.CompileError> generate(magma.compile.Node node){return left.generate(node).and(__lambda0__.generate(node)).mapValue(__lambda1__.left()+infix+tuple.right());
}
